pxtcompiler/emitter/service.ts

```diff
@@ -1723,22 +1734,61 @@ namespace ts.pxtc.service {
 . . . . .
 """`;
 
-    interface SnippetResult {
-        snippet: string,
-        addsDefinitions: boolean
+    export type SnippetNode = SnippetReplacePoint | string | SnippetNode[]
+    export interface SnippetReplacePoint {
+        default: SnippetNode;
+        isDefinition?: boolean;
     }
-    function asSnippetRes(snippet: string) {
-        return {
-            snippet,
-            addsDefinitions: false
+    function isSnippetReplacePoint(n: SnippetNode): n is SnippetReplacePoint {
+        return !!(n as SnippetReplacePoint).default
+    }
+    function isSnippetNodeList(n: SnippetNode): n is SnippetNode[] {
+        return typeof (n) === "object" && typeof ((n as SnippetNode[]).length) === "number"
+    }
+    export function snippetStringify(snippet: SnippetNode, emitMonacoReplacementPoints = false): string {
+        // The format for monaco snippets is: 
+        //      foo(${1:bar}, ${2:baz},  ${1:bar})
```

Comment:
Hm, so is the multicursor behavior for same type parameters intended? I always found that behavior a bit confusing personally

Reply:
Multi cursor should only be for the same identifiers. I have a bug right now that hit happens for the same types like "false" but I'll fix that.